numpyro/contrib/einstein/reinit_guide.py

```diff
@@ -0,0 +1,77 @@
+# Copyright Contributors to the Pyro project.
+# SPDX-License-Identifier: Apache-2.0
+
+from abc import ABC, abstractmethod
+
+import jax.random
+
+from numpyro import handlers
+from numpyro.distributions.constraints import real
+from numpyro.infer.util import init_to_uniform
+
+__all__ = ["ReinitGuide", "WrappedGuide"]
+
+
+class ReinitGuide(ABC):
+    @abstractmethod
+    def init_params(self):
+        raise NotImplementedError
+
+    @abstractmethod
+    def find_params(self, rng_keys, *args, **kwargs):
+        raise NotImplementedError
+
+
+class WrappedGuide(ReinitGuide):
+    def __init__(
+        self,
+        fn,
+        reinit_hide_fn=lambda site: site["name"].endswith("$params"),
+        init_strategy=init_to_uniform,
+    ):
+        self.fn = fn
+        self._init_params = None
+        self.init_strategy = init_strategy(
+            reinit_param=lambda site: not reinit_hide_fn(site)
+        )
+        self._reinit_hide_fn = reinit_hide_fn
+
+    def init_params(self):
+        return self._init_params
+
+    def find_params(self, rng_keys, *args, **kwargs):
+        def _find_valid_params(rng_key):
+            trace_key, site_key = jax.random.split(rng_key)
+            guide_trace = handlers.trace(handlers.seed(self.fn, rng_key)).get_trace(
+                *args, **kwargs
+            )
+            params = {
+                name: site["value"]
+                for name, site in guide_trace.items()
+                if site["type"] == "param" and self._reinit_hide_fn(site)
+            }
+            for site in guide_trace.values():
+                if site["type"] == "param" and not self._reinit_hide_fn(site):
+                    site_key, rng_key = jax.random.split(site_key)
+                    site["kwargs"]["rng_key"] = rng_key
```

Comment:
Another way to not touch init strategies is to change the site type, fn, is observed keys for this site. (You can use ImproperUniform for `fn`.)

Reply:
I don't particularly like the `reinit_hide_fn` as it requires the user to list parameters twice. A solution that eliminates it is superior, in my opinion.